src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -122,6 +122,42 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                     self.binop_with_overflow(bin_op, lhs, rhs, dest)?;
                 }
             }
+            "saturating_add" | "saturating_sub" => {
+                let l = self.read_immediate(args[0])?;
+                let r = self.read_immediate(args[1])?;
+                let is_add = intrinsic_name == "saturating_add";
+                let (val, overflowed) = self.binary_op_imm(if is_add {
+                    BinOp::Add
+                } else {
+                    BinOp::Sub
+                }, l, r)?;
+                let val = if overflowed {
+                    // For signed ints the saturated value depends on the
+                    // sign of the first term
+                    let first_term: u128 = l.to_scalar()?.to_bits(l.layout.size)?;
+                    let num_bits = l.layout.size.bits();
+                    if l.layout.abi.is_signed() {
+                        if first_term & (1 << (num_bits-1)) == 0 {  // first term is positive
```

Comment:
So it cannot overflow on the negative side because the smallest possible things we could reach here (for u8, similar for other types) are "0 - 127" and "0 + -128", and both are fine.  Correct?